src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -158,106 +158,106 @@ AcquireMetadataLocks(List *taskList)
-		 * The executor shard lock is used to maintain consistency between
-		 * replicas and therefore no lock is required for read-only queries
-		 * or in general when there is only one replica.
-		 */
+		LOCKMODE lockMode = NoLock;
 
-		lockMode = NoLock;
-	}
-	else if (list_length(task->taskPlacementList) == 1)
-	{
-		if (task->replicationModel == REPLICATION_MODEL_2PC)
+		if (modLevel <= MODLEVEL_READONLY)
 		{
 			/*
-			 * While we don't need a lock to ensure writes are applied in
-			 * a consistent order when there is a single replica. We also use
-			 * shard resource locks as a crude implementation of SELECT..
-			 * FOR UPDATE on reference tables, so we should always take
-			 * a lock that conflicts with the FOR UPDATE/SHARE locks.
+			 * The executor shard lock is used to maintain consistency between
+			 * replicas and therefore no lock is required for read-only queries
+			 * or in general when there is only one replica.
 			 */
-			lockMode = RowExclusiveLock;
+
+			lockMode = NoLock;
 		}
-		else
+		else if (list_length(task->taskPlacementList) == 1)
+		{
+			if (task->replicationModel == REPLICATION_MODEL_2PC)
+			{
+				/*
+				 * While we don't need a lock to ensure writes are applied in
+				 * a consistent order when there is a single replica. We also use
+				 * shard resource locks as a crude implementation of SELECT..
+				 * FOR UPDATE on reference tables, so we should always take
+				 * a lock that conflicts with the FOR UPDATE/SHARE locks.
+				 */
+				lockMode = RowExclusiveLock;
+			}
+			else
+			{
+				/*
+				 * When there is no replication, the worker itself can decide on
+				 * on the order in which writes are applied.
+				 */
+				lockMode = NoLock;
+			}
+		}
+		else if (AllModificationsCommutative)
 		{
 			/*
-			 * When there is no replication, the worker itself can decide on
-			 * on the order in which writes are applied.
+			 * Bypass commutativity checks when citus.all_modifications_commutative
+			 * is enabled.
+			 *
+			 * A RowExclusiveLock does not conflict with itself and therefore allows
+			 * multiple commutative commands to proceed concurrently. It does
+			 * conflict with ExclusiveLock, which may still be obtained by another
+			 * session that executes an UPDATE/DELETE/UPSERT command with
+			 * citus.all_modifications_commutative disabled.
 			 */
-			lockMode = NoLock;
-		}
-	}
-	else if (AllModificationsCommutative)
-	{
-		/*
-		 * Bypass commutativity checks when citus.all_modifications_commutative
-		 * is enabled.
-		 *
-		 * A RowExclusiveLock does not conflict with itself and therefore allows
-		 * multiple commutative commands to proceed concurrently. It does
-		 * conflict with ExclusiveLock, which may still be obtained by another
-		 * session that executes an UPDATE/DELETE/UPSERT command with
-		 * citus.all_modifications_commutative disabled.
-		 */
 
-		lockMode = RowExclusiveLock;
-	}
-	else if (task->upsertQuery || commandType == CMD_UPDATE || commandType == CMD_DELETE)
```

Comment:
Need to make sure `modLevel == MODLEVEL_NONCOMMUTATIVE` when `task->upsertQuery`

Reply:
Hmz, seems like we're missing some test coverage.